File: model-endpoint/guided-json.py
#!/usr/bin/env python
"""
Example of using guided JSON output with our API.
This demonstrates how to get structured data from the model using a JSON schema.
"""

import os
import json
from openai import OpenAI
from dotenv import load_dotenv

# Sample JSON schema to guide the model's output
# Note: This is a simple example - schema capabilities have some limitations
SAMPLE_SCHEMA = {
  "type": "object",
  "properties": {
    "analysis": {
      "type": "object",
      "properties": {
        "title": {"type": "string"},
        "summary": {"type": "string"},
        "key_points": {
          "type": "array",
          "items": {"type": "string"}
        },
        "sentiment": {
          "type": "string",
          "enum": ["positive", "neutral", "negative"]
        },
        "categories": {
          "type": "array",
          "items": {"type": "string"}
        }
      },
      "required": ["summary", "key_points", "sentiment"]
    }
  },
  "required": ["analysis"]
}

def main():
    # Load environment variables from .env file
    load_dotenv()
    
    # Get API credentials from environment variables
    base_url = os.getenv("BASE_URL")
    api_key = os.getenv("API_KEY")
    model = os.getenv("MODEL_NAME")
    
    # Verify credentials are available
    if not base_url or not api_key:
        print("Error: BASE_URL and API_KEY must be set in .env file")
        return
    
    # Ensure base_url ends with /v1
    if not base_url.endswith('/v1'):
        base_url = base_url.rstrip('/') + '/v1'
    
    # Initialize the OpenAI client with our API endpoint
    client = OpenAI(
        base_url=base_url,
        api_key=api_key
    )
    
    # Prepare user prompt that will work well with the JSON schema
    messages = [
        {"role": "system", "content": "You are a helpful assistant that provides analysis in structured JSON format."},
        {"role": "user", "content": "Analyze the following text and provide structured insights: "
         "Renewable energy investments hit a record high last quarter, with solar projects "
         "leading the way. Despite supply chain challenges, new installations increased by 28%. "
         "However, regulatory uncertainty in some markets has caused concern among investors."}
    ]
    
    # Create the API request with guided JSON output
    print("Requesting guided JSON response...")
    try:
        # Note: The extra_body parameter is used to pass the JSON schema
        response = client.chat.completions.create(
            model=model,
            messages=messages,
            temperature=0.5,
            max_tokens=500,
            extra_body={
                "guided_json": SAMPLE_SCHEMA,
                "guided_decoding_backend": "xgrammar"  # This backend is used for decoding
            }
        )
        
        # Extract and parse the response
        json_response = response.choices[0].message.content
        
        # Display the raw JSON response
        print("\nRaw JSON Response:")
        print("-" * 60)
        print(json_response)
        print("-" * 60)
        
        # Parse and pretty-print the JSON
        try:
            parsed_json = json.loads(json_response)
            print("\nPretty-printed JSON:")
            print("-" * 60)
            print(json.dumps(parsed_json, indent=2))
            print("-" * 60)
            
            # Example of accessing specific fields
            if "analysis" in parsed_json:
                analysis = parsed_json["analysis"]
                print("\nHighlights from the analysis:")
                print(f"Sentiment: {analysis.get('sentiment', 'N/A')}")
                print(f"Summary: {analysis.get('summary', 'N/A')}")
                
                if "key_points" in analysis:
                    print("\nKey Points:")
                    for i, point in enumerate(analysis["key_points"], 1):
                        print(f"{i}. {point}")
        except json.JSONDecodeError:
            print("Error: The response is not valid JSON")
        
    except Exception as e:
        print(f"Error: {e}")
        
    print("\nNote: Our guided JSON capabilities have some limitations.")
    print("For complex schemas or special requirements, please contact our support team.")

if __name__ == "__main__":
    main()

File: model-endpoint-ts/guided-json.ts
#!/usr/bin/env ts-node
/**
 * Example of using guided JSON output with our API.
 * This demonstrates how to get structured data from the model using a JSON schema.
 */

import * as dotenv from 'dotenv';
import OpenAI from 'openai';

// Sample JSON schema to guide the model's output
// Note: This is a simple example - schema capabilities have some limitations
// Note: Since vLLM v0.8.3 enums are supported in the schema. If using vLLM >= v0.8.3, you can
// modify the sentiment field to be an enum with values like "positive", "negative", "neutral"
// and the model will respect that. Like this:
// "sentiment": {
//   "type": "string",
//   "enum": ["positive", "negative", "neutral"]
// }
const SAMPLE_SCHEMA = {
  type: 'object',
  properties: {
    analysis: {
      type: 'object',
      properties: {
        title: { type: 'string' },
        summary: { type: 'string' },
        key_points: {
          type: 'array',
          items: { type: 'string' }
        },
        sentiment: {
          type: 'string'
        },
        categories: {
          type: 'array',
          items: { type: 'string' }
        }
      },
      required: ['summary', 'key_points', 'sentiment']
    }
  },
  required: ['analysis']
};

interface AnalysisResponse {
  analysis: {
    title?: string;
    summary: string;
    key_points: string[];
    sentiment: string;
    categories?: string[];
  };
}

async function main(): Promise<void> {
  // Load environment variables from .env file
  dotenv.config();
  
  // Get API credentials from environment variables
  const baseUrl = process.env.BASE_URL;
  const apiKey = process.env.API_KEY;
  const model = process.env.MODEL_NAME;
  
  // Verify credentials are available
  if (!baseUrl || !apiKey) {
    console.error('Error: BASE_URL and API_KEY must be set in .env file');
    return;
  }
  
  // Ensure baseUrl ends with /v1
  const baseUrlNormalized = !baseUrl.endsWith('/v1') 
    ? baseUrl.replace(/\/+$/, '') + '/v1'
    : baseUrl;
  
  // Initialize the OpenAI client with our API endpoint
  const client = new OpenAI({
    baseURL: baseUrlNormalized,
    apiKey: apiKey
  });
  
  // Prepare user prompt that will work well with the JSON schema
  const messages = [
    { role: 'system', content: 'You are a helpful assistant that provides analysis in structured JSON format.' },
    { role: 'user', content: 'Analyze the following text and provide structured insights: ' +
      'Renewable energy investments hit a record high last quarter, with solar projects ' +
      'leading the way. Despite supply chain challenges, new installations increased by 28%. ' +
      'However, regulatory uncertainty in some markets has caused concern among investors.' }
  ] as OpenAI.ChatCompletionMessageParam[];
  
  // Create the API request with guided JSON output
  console.log('Requesting guided JSON response...');
  try {
    // Note: We pass the JSON schema via the extra/extra_body parameter
    const response = await client.chat.completions.create({
      model: model || '',
      messages: messages,
      temperature: 0.5,
      max_tokens: 500,
      extra: {
        guided_json: SAMPLE_SCHEMA,
        guided_decoding_backend: 'xgrammar'  // This backend is used for decoding
      }
    } as any); // Type casting for extra parameters
    
    // Extract the response
    const jsonResponse = response.choices[0].message.content;
    
    // Display the raw JSON response
    console.log('\nRaw JSON Response:');
    console.log('-'.repeat(60));
    console.log(jsonResponse);
    console.log('-'.repeat(60));
    
    // Parse and pretty-print the JSON
    try {
      const parsedJson = JSON.parse(jsonResponse || '{}') as AnalysisResponse;
      console.log('\nPretty-printed JSON:');
      console.log('-'.repeat(60));
      console.log(JSON.stringify(parsedJson, null, 2));
      console.log('-'.repeat(60));
      
      // Example of accessing specific fields
      if (parsedJson.analysis) {
        const analysis = parsedJson.analysis;
        console.log('\nHighlights from the analysis:');
        console.log(`Sentiment: ${analysis.sentiment || 'N/A'}`);
        console.log(`Summary: ${analysis.summary || 'N/A'}`);
        
        if (analysis.key_points && analysis.key_points.length > 0) {
          console.log('\nKey Points:');
          analysis.key_points.forEach((point, i) => {
            console.log(`${i + 1}. ${point}`);
          });
        }
      }
    } catch (error) {
      console.error('Error: The response is not valid JSON');
    }
    
  } catch (error: any) {
    console.error(`Error: ${error.message || error}`);
  }
  
  console.log('\nNote: Our guided JSON capabilities have some limitations.');
  console.log('For complex schemas or special requirements, please contact our support team.');
}

if (require.main === module) {
  main().catch(console.error);
}